### ISSUE TITLE
A strange training stuck

### ISSUE THREAD

**Issue by Remember2018** [tensorpack/tensorpack#1134]
Hi, I'm using the `FasterRCNN` module in Tensorpack. And I modified the original network structures (`ResNetC4Model` and `ResNetFPNModel`) to fit my problem (simply add some other branchs). Training both networks with one GPU are OK. And training ResNetC4Model with two GPUs is also OK. But training ResNetFPNModel with two GPUs would fail. The training seems stuck in the `GPUUtilizationTracker` for a long time and no `python` process or `ZMQbg` processes can be found in `top` command during the stuck. The screenshot is 
![1](https://user-images.githubusercontent.com/43327429/55779037-13502a00-5ad7-11e9-9d0f-927472f190b9.png)
But when I comment the following two lines in `train.py`,

        # if not is_horovod:
        #     callbacks.append(GPUUtilizationTracker())

the training would be OK, 
![2](https://user-images.githubusercontent.com/43327429/55779216-72ae3a00-5ad7-11e9-8008-d4b0f1c35e33.png)
I'm not sure why this happened. And this may be difficult for you to reproduce or diagnose. I can confirm the two line comments woule solve the stuck. If convenient, please have a check with the `GPUUtilizationTracker`, thanks.

**Comment by ppwwyyxx** [comment]
This sounds like a bug. For anyone to better understand and diagnose your issue, please post relevant details following the issue template (you can find it by clicking "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))

Also, please confirm whether you can reliably reproduce this bug every time or not, and whether the workaround you've found is effective every time or not..

**Comment by ppwwyyxx** [comment]
In addition to the above requests:
1. Could you reproduce this issue without modification to the model?
2. When it is stuck what would you observe if you press `ctrl-C`? (include logs if any)

**Comment by Remember2018** [comment]
Hi @ppwwyyxx , thanks for your suggestions. I have made a video recording the training process. Please find the video in https://pan.baidu.com/s/1PcJW4Aapt1W3iRGwcAfQ-w, the extraction code is `7kqr`. 

In this video, I run the experiments `without` and `with` the two line comment. 
First `without` the two line comment (from 0:00 to 9:40), please pay more attention the video from 2:10 to 4:10, which shows a short term stuck during the training, and after the first epoch, the training will stuck.

Then I comment the two line (from 9:40 to end), the training is OK. 

Please note that the experiments are with my modified code based on ResNetC4Model and ResNetFPNModel. I think the cause of the stuck lies in my modification to the ResNetFPNModel. Please allow me to check it again and check it carefully! Training ResNetFPNModel(modified) with one/two GPUs will have the stuck.

1. Could you reproduce this issue without modification to the model?
No. The original model run OK.

2. When it is stuck what would you observe if you press ctrl-C? (include logs if any)
Please see the video if convenient. It shows "2019-04-09 17:24:03.079834: W tensorflow/core/kernels/queue_base.cc:277] _0_QueueInput/input_queue: Skipping cancelled enqueue attempt with queue not closed"

**Comment by ppwwyyxx** [comment]
Thanks for the video. It seems there is an error in the GPUUtilizationTracker worker process but the error was somehow not printed. Could you uninstall and reinstall the [latest tensorpack](https://github.com/tensorpack/tensorpack/commit/6e0b509b5a010ab1ca14033999fa75cef39d579c) and see whether there are any error logs? If no extra information is printed, you might need to add some print statements in `GPUUtilizationTracker.worker` to see where it is stuck.

**Comment by Remember2018** [comment]
@ppwwyyxx Thanks for checking this problem! I've installed the latest master and also add the following in the `prof.py`:
![2](https://user-images.githubusercontent.com/43327429/55848642-2d921280-5b80-11e9-93d4-e354ad8fee66.jpg)

Then the log is as follows:
![1](https://user-images.githubusercontent.com/43327429/55848750-9e392f00-5b80-11e9-9535-6c95680ab559.jpg)

If convenient, please have a check. Thanks!

**Comment by Remember2018** [comment]
@ppwwyyxx And I further add more log in `prof.py` as follows:
![3](https://user-images.githubusercontent.com/43327429/55849267-97abb700-5b82-11e9-8005-39b144d687ec.jpg)
then the log is as follows:
![4](https://user-images.githubusercontent.com/43327429/55849303-b447ef00-5b82-11e9-8303-4b2daf1a0b56.jpg)

**Comment by Remember2018** [comment]
@ppwwyyxx I've attached the logs. Please kindly check them. Thanks.
[c4_log.log](https://github.com/tensorpack/tensorpack/files/3061884/c4_log.log)
[fpn_log.log](https://github.com/tensorpack/tensorpack/files/3061885/fpn_log.log)

**Comment by ppwwyyxx** [comment]
Thanks for the log. These are helpful.
As you can see, your FPN log was stuck between "worker 5" and "worker 6" however the C4 log was not.

This is quite strange, given that the code between them is essentially equivalent to this:
```python
import time
import numpy as np
from tensorpack.utils.nvml import NVMLContext

stats = np.zeros((2,), dtype='f4')
cnt = 0
with NVMLContext() as ctx:
    while True:
        time.sleep(1)
        data = [ctx.device(i).utilization()['gpu'] for i in [0, 1]]
        data = list(map(float, data))
        stats += data
        cnt += 1
        print(data)
```
which should never get stuck. It makes call to the nvidia driver, so if it ever gets stuck it sounds like a bug in the driver.

**Comment by ppwwyyxx** [comment]
The latest commit (f831a46e7f37b1bf08b65e1f3549a8ee4f9921b2) reorganizes the above logic to make fewer calls to nvidia driver. It will only make calls to `utilization()` but not `device()`. It is a shot in the dark but if there is indeed an issue in the driver, this is something worth trying.

Other than this I don't have other suggestions except to upgrade the driver.

**Comment by Remember2018** [comment]
Hi @ppwwyyxx , thanks a lot for your suggestions! 
I tried to add more prints in `prof.py` and the log is attached. Please note in the video above, there is  a short term stuck from 2:10 to 4:10. It seems after this short term stuck, the `self._proc.is_alive()` seems to be `False`. I think the later stuck is much realted to this. But I cannot figure out why the `self._proc` is stopped during that short term stuck.
![5](https://user-images.githubusercontent.com/43327429/55860742-81fcb880-5ba7-11e9-88cf-1d3ef937a0f3.jpg)

Please find the detail `prof.py` and `log.log` as follows:
[log.log](https://github.com/tensorpack/tensorpack/files/3062565/log.log)

[prof.py.txt](https://github.com/tensorpack/tensorpack/files/3062576/prof.py.txt)


I'll update Tensorpack with your latest commit and test. Thanks!

**Comment by ppwwyyxx** [comment]
Thanks. It seems that this time it's stuck (or died) after "worker 6", different from last time. So perhaps it's not stuck in the nvidia driver and may be the cause is in somewhere else.

However this still feels very strange. It does seem like the process died rather than stuck.

You could take a look at `dmesg` command and see if it has printed any information about processes being killed by the operating system.

Perhaps also try removing the line `ensure_proc_terminate(self._proc)` from `_before_train` and see if it changes anything?

**Comment by Remember2018** [comment]
I have update the `prof.py` with your latest commit. And check the outputs of `dmesg`.
Please find the `dmesg.txt`, it seems the `ZMQ` process was killed due to `Out of Memory`. 
![7](https://user-images.githubusercontent.com/43327429/55864109-66e17700-5bae-11e9-97ac-9de1f0f2a8a6.jpg)
![6](https://user-images.githubusercontent.com/43327429/55864118-6c3ec180-5bae-11e9-8689-2b447fd6c121.jpg)
[prof.py_new.txt](https://github.com/tensorpack/tensorpack/files/3062799/prof.py_new.txt)
[dmesg.txt](https://github.com/tensorpack/tensorpack/files/3062811/dmesg.txt)

@ppwwyyxx Thank you very much for helping checking this consistently! 
I'm sorry but I will be out for a few hours. I'll update this issue as soon as possible. Thanks.

**Comment by Remember2018** [comment]
@ppwwyyxx After I comment the two lines related to `GPUUtilizationTracker` in `train.py`, then the training is OK. And the memory usage is also high, but no stuck. So this may be not the main reason.
![8](https://user-images.githubusercontent.com/43327429/55883147-5e059b00-5bd8-11e9-9134-d5a0213b7015.jpg)

**Comment by ppwwyyxx** [comment]
Thanks for the investigation. So indeed some processes are killed. Not sure whether it is the `GPUUtilizationTracker` process -- but in any case a killed process may stuck your training.

I'm not sure why no processes are killed when `GPUUtilizationTracker` is not used, despite similar memory usage. I have some hypothesis but I'll need some time to try it.

For now, you need to lower your memory usage. You can achieve this by using fewer processes or smaller `buffer_size` in https://github.com/tensorpack/tensorpack/blob/8dd254be676d89000a7f4f6fd5cebe78b3499789/examples/FasterRCNN/data.py#L375.

**Comment by Remember2018** [comment]
Hi @ppwwyyxx , the problem lies in the memory  and my wrong setting. In my previous `config.py`: 

        _C.PREPROC.TRAIN_SHORT_EDGE_SIZE = [1024, 1024]  # [min, max] to sample from
        _C.PREPROC.TEST_SHORT_EDGE_SIZE = 1024
        _C.PREPROC.MAX_SIZE = 2048

which is the cause of the `out of memory`. 

And after I set the `config.py` as follows:

        _C.PREPROC.TRAIN_SHORT_EDGE_SIZE = [800, 800]  # [min, max] to sample from
        _C.PREPROC.TEST_SHORT_EDGE_SIZE = 800
        _C.PREPROC.MAX_SIZE = 1024

The problem gone. It seems the `Out of memory` will have effects on the internal process in `GPUUtilizationTracker` class and terminate it during training. Not sure why this happened.

And please check the following line: https://github.com/tensorpack/tensorpack/blob/6e0b509b5a010ab1ca14033999fa75cef39d579c/tensorpack/callbacks/prof.py#L78

The `stats` seems a list, so you may improve that. Thanks a lot!

**Comment by ppwwyyxx** [comment]
The root cause of this issue is that the host runs out-of-memory and a worker process is forced killed by OS, causing the stuck.
The reason why `GPUUtilizationTracker` leads to increased memory use is due to how `fork` works - it duplicates the current process and therefore may increase memory consumption.

We have added a few improvements in recent diffs:
1. A `HostMemoryTracker` callback to track memory consumption during training.
2. fork early in `GPUUtilizationTracker` to save some memory
3. Use less memory in the data loader.

A more fundamental solution to reduce memory usage is to use the `forkserver/spawn` context in Python's `multiprocessing`, available since Python 3.4.
